File: C#_Base_PT4/OOP3Behav4.cs
// File: "OOP3Behav4"
using PT4;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace PT4Tasks
{
    public class MyTask : PT
    {
        public class Validator
        {
            public virtual string Validate(string s)
            {
                return "";
            }
        }

        // Implement the EmptyValidator, NumberValidator
        //   and RangeValidator descendant classes
        /////
        ///

        public class EmptyValidator : Validator
        {
            public EmptyValidator(){}
            public override string Validate(string s)
            {
                if(string.IsNullOrEmpty(s)) return "!Empty text";
                return "";
            }
        }

        public class NumberValidator : Validator
        {
            public NumberValidator(){}
            public override string Validate(string s)
            {
                string str = "";
                if(s != "")
                {
                    int k = 0;
                    for(int i = 0; i < s.Length; i++)
                    {
                        if(s[i] != '-' && (s[i] < '0' || s[i] > '9'))
                        {
                            k = 1;
                            break;
                        }
                    }
                    if(k == 0) str = "";
                    else str = "!'" + s + "': not a number";
                }
                else str = "!'" + s + "': not a number";
                return str;
            }
        }

        public class RangeValidator : Validator
        {
            int my_min, my_max;
            public RangeValidator(){}
            public RangeValidator(int a, int b)
            {
                this.my_min = Math.Min(a,b);
                this.my_max = Math.Max(a,b);
            }
            public override string Validate(string s)
            {
                string str = "";
                if(s != "")
                {
                    int k = 0;
                    for(int i = 0; i < s.Length; i++)
                    {
                        if(s[i] != '-' && (s[i] < '0' || s[i] > '9'))
                        {
                            k = 1;
                            break;
                        }
                    }
                    if(k == 0)
                    {
                        int x = int.Parse(s);
                        if ((x >= my_min) && (x <= my_max)) 
                            str = "";
                        else 
                            str = "!'" + s + "': not in range " + Convert.ToString(my_min) + ".." + Convert.ToString(my_max);
                    }
                    else{
                        str = "!'" + s + "': not in range " + Convert.ToString(my_min) + ".." + Convert.ToString(my_max);
                    }
                }
                else{
                    str = "!'" + s + "': not in range " + Convert.ToString(my_min) + ".." + Convert.ToString(my_max);
                }         
                return str;
            }
        }

        // Implement the TextBox and TextForm classes

        public class TextBox
        {
            string text = "";
            Validator v = new Validator();
            public TextBox(){}
            public void SetText(string text)
            {
                this.text = text;
            }
            public void SetValidator(Validator v)
            {
                this.v = v;
            }
            public string Validate()
            {
                return v.Validate(text);
            }
        }

        public class TextForm 
        {
            List<TextBox> V = new List<TextBox>();
            string info = "";
            public TextForm(int m)
            {
                for(int i = 0; i < m; i++)
                {
                    TextBox tb = new TextBox();
                    V.Add(tb);
                }
            }
            public void SetText(int ind, string text)
            {
                V[ind].SetText(text);
            }

            public void SetValidator (int ind, Validator v)
            {
                V[ind].SetValidator(v);
            }

            public string Validate()
            {
                foreach(TextBox i in V)
                {
                    info = info + i.Validate();
                }
                return info;
            }
        }
        public static void Solve()
        {
            Task("OOP3Behav4");
            int N = GetInt(),
                A = GetInt(),
                B = GetInt(),
                K = GetInt();

            int[] v1 = new int [K];
            char[] v2 = new char [K]; 

            for (int i = 0; i < K; i++) {
                int ind = GetInt();
                char val = GetChar();
                v1[i] = ind;
                v2[i] = val;
            }

            for(int i = 0; i < 5; i++)
            {
                TextForm tf = new TextForm (N);
                for(int j = 0; j < N; j++)
                {
                    string s = GetString();
                    tf.SetText(j, s);
                }
                Validator v = new Validator();
                for(int k = 0; k < K; k++)
                {
                    if(v2[k] == 'E') 
                    {
                        Show("empty");
                        v = new EmptyValidator();
                    }
                    if(v2[k] == 'N') 
                    {
                        v = new NumberValidator();
                        Show("number");
                    }
                    if(v2[k] == 'R') 
                    {
                        v = new RangeValidator(A, B);
                        Show("range");
                    }
                    
                    tf.SetValidator(v1[k], v);
                }
                 
                ShowLine(tf.Validate());
                string str = tf.Validate();
                string info = "";
                for(int n = 0; n < str.Length/2; n++)
                {
                    info = info + str[n];
                }
                Put(info);
                // ShowLine(info.Length);
            }
        }
    }
}

File: C++_Base_PT4/OOP3Behav4.cpp
#include "pt4.h"
#include <vector>
#include <algorithm>

using namespace std;

class Validator
{
public:
    virtual string Validate(string s);
};

string Validator::Validate(string s)
{
	return "";
}

// §²§¦§¡§­§ª§©§µ§«§´§¦ §¬§­§¡§³§³§½-§±§°§´§°§®§¬§ª EmptyValidator,
//   NumberValidator §ª RangeValidator

class EmptyValidator: public Validator
{
public:
	EmptyValidator() {}
	string Validate(string s)
	{
		if (s.empty()) return "!Empty text";
		return "";
	}
};

class NumberValidator: public Validator
{
public:
	NumberValidator() {}
	string Validate(string s)
	{
		string str;
		if(s != "")
		{
			string::iterator p1 = find_if(s.begin(), s.end(), [](char c) { return (c != '-') && ((c < '0') || (c > '9')); });
			if (p1 == s.end()) str = "";
			else str = "!'" + s + "': not a number";
		}
		else str = "!'" + s + "': not a number";
		return str;
	}
};

class RangeValidator: public Validator
{
public:
	RangeValidator() {}
	RangeValidator(int a, int b)
	{
		this->my_min = min(a, b);
		this->my_max = max(a, b);
	}
	string Validate(string s)
	{
		string str;
		string::iterator p1 = find_if(s.begin(), s.end(), [](char c) { return (c != '-') && ((c < '0') || (c > '9')); });
		if (p1 == s.end()) 
		{
			int x = atoi(s.c_str());
			if ((x >= my_min) && (x <= my_max)) 
				str = "";
			else 
				str = "!'" + s + "': not in range " + to_string(my_min) + ".." + to_string(my_max);
		}
		else 
			str = "!'" + s + "': not in range " + to_string(my_min) + ".." + to_string(my_max);
		return str;
	}
private:
	int my_min, my_max;
};

// §²§¦§¡§­§ª§©§µ§«§´§¦ §¬§­§¡§³§³§½ TextBox §ª TextForm

class TextBox
{
public:
	TextBox() {}
	~TextBox()
	{
		delete v;
		v = NULL;
	}
	void SetText(string text)
	{
		this->text = text;
	}
	void SetValidator(Validator *v)
	{
		this->v = v;
	}
	string Validate()
	{
		return v->Validate(text);
	}
private:
	string text = "";
	Validator *v = new Validator();
};

class TextForm
{	
public:
	TextForm(int n)
	{
		for (int i = 0; i < n; i++) {
			TextBox* tb = new TextBox();
			V.push_back(tb);
		}
	}

	void SetText(int ind,string text)
	{
		V[ind]->SetText(text);
	}

	void SetValidator (int ind,Validator * v)
	{
		V[ind]->SetValidator(v);
	}

	string Validate()
	{
		for(auto i : V)
		{
			info = info + i->Validate();
		}
		return info;
	}
private:
	vector<TextBox*> V;
	string info = "";
};

void Solve()
{
    Task("OOP3Behav4");
	int N, A, B, K;
	pt >> N >> A >> B;
	pt >> K;

	vector<pair<int, char>> V;
	for (int i = 0; i < K; i++) {
		int ind;
		char val;
		pt >> ind >> val;
		V.emplace_back(make_pair(ind, val));
	}
	
	for (int i = 0; i < 5; i++) 
	{
		TextForm* tf = new TextForm(N);
		for (int j = 0; j < N; j++) {
			string s;
			pt >> s;
			tf->SetText(j, s);
		}
		Validator* v = NULL;
		for (auto e : V) {
			if (e.second == 'E') v = new EmptyValidator();
			if (e.second == 'N') v = new NumberValidator();
			if (e.second == 'R') v = new RangeValidator(A, B);
			tf->SetValidator(e.first, v);
		}
		ShowLine(tf->Validate());
		string str = tf->Validate();
		string info;
		for(int i = 0; i < str.size()/2; i++)
		{
			info = info + str[i];
		}
	//	pt << str; 
	ShowLine(info); 
	ShowLine(info.size()); 
	pt << info;
	}
}
